File: src/main/java/cc/hyperium/launch/patching/conflicts/GameSettingsTransformer.java
package cc.hyperium.launch.patching.conflicts;

import org.objectweb.asm.Label;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.*;

import java.util.Iterator;

public class GameSettingsTransformer implements ConflictTransformer {
    @Override
    public String getClassName() {
        return "net.minecraft.client.settings.GameSettings";
    }

    @Override
    public ClassNode transform(ClassNode original) {
        original.visitField(Opcodes.ACC_PRIVATE, "needsResourceRefresh", "Z", null, null).visitEnd();
        MethodNode node = new MethodNode(Opcodes.ACC_PUBLIC, "onGuiClosed", "()V", null, null);
        node.maxLocals = 1;
        node.maxStack = 2;
        node.instructions.add(createOnGuiClosedList());
        original.methods.add(node);
        for (MethodNode method : original.methods) {
            if (method.name.equals("setOptionFloatValue")) {
                // needsResourceRefresh = true; return;
                InsnList l = new InsnList();
                l.add(new VarInsnNode(Opcodes.ALOAD, 0));
                l.add(new InsnNode(Opcodes.ICONST_1));
                l.add(new FieldInsnNode(Opcodes.PUTFIELD, "net/minecraft/client/settings/GameSettings", "needsResourceRefresh", "Z"));
                l.add(new InsnNode(Opcodes.RETURN));
                // insert it after setBlurMipmapDirect
                Iterator<AbstractInsnNode> it = method.instructions.iterator();
                while (it.hasNext()) {
                    AbstractInsnNode insn = it.next();
                    if (insn instanceof MethodInsnNode && insn.getOpcode() == Opcodes.INVOKEVIRTUAL
                            && ((MethodInsnNode) insn).owner.equals("net/minecraft/client/Minecraft")
                            && ((MethodInsnNode) insn).name.equals("scheduleResourcesRefresh")
                            && ((MethodInsnNode) insn).desc.equals("()Lcom/google/common/util/concurrent/ListenableFuture;")) {
                        method.instructions.insertBefore(insn.getPrevious().getPrevious(), l);
                        break;
                    }
                }
                method.instructions.insert(new MethodInsnNode(Opcodes.INVOKEVIRTUAL,
                        "net/minecraft/client/renderer/texture/TextureMap",
                        "setBlurMipmapDirect",
                        "(ZZ)V",
                        false), l);
            }
        }
        return original;
    }

    private InsnList createOnGuiClosedList() {
        InsnList l = new InsnList();
        LabelNode label = new LabelNode(new Label());
        // if (this.needsResourceRefresh)
        l.add(new VarInsnNode(Opcodes.ALOAD, 0));
        l.add(new FieldInsnNode(Opcodes.GETFIELD, "net/minecraft/client/settings/GameSettings", "needsResourceRefresh", "Z"));
        l.add(new JumpInsnNode(Opcodes.IFEQ, label));
        // this.mc.scheduleResourcesRefresh()
        l.add(new VarInsnNode(Opcodes.ALOAD, 0));
        l.add(new FieldInsnNode(Opcodes.GETFIELD, "net/minecraft/client/settings/GameSettings", "mc", "Lnet/minecraft/client/Minecraft;")); // this.mc
        l.add(new MethodInsnNode(Opcodes.INVOKEVIRTUAL, "net/minecraft/client/Minecraft", "scheduleResourcesRefresh",
                "()Lcom/google/common/util/concurrent/ListenableFuture;", false)); // scheduleResourcesRefresh()
        l.add(new InsnNode(Opcodes.POP));
        // this.needsResourceRefresh = false
        l.add(new VarInsnNode(Opcodes.ALOAD, 0));
        l.add(new InsnNode(Opcodes.ICONST_2));
        l.add(new FieldInsnNode(Opcodes.PUTFIELD, "net/minecraft/client/settings/GameSettings", "needsResourceRefresh", "Z"));
        // return
        l.add(label);
        l.add(new InsnNode(Opcodes.RETURN));
        return l;
    }
}

File: src/main/kotlin/cc/hyperium/launch/patching/conflicts/GameSettingsTransformer.kt
package cc.hyperium.launch.patching.conflicts

import codes.som.anthony.koffee.assembleBlock
import codes.som.anthony.koffee.insns.jvm.*
import codes.som.anthony.koffee.koffee
import com.google.common.util.concurrent.ListenableFuture
import net.minecraft.client.Minecraft
import net.minecraft.client.settings.GameSettings
import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.*

class GameSettingsTransformer : ConflictTransformer {
    override fun getClassName() = "avh"

    override fun transform(original: ClassNode): ClassNode {
        original.visitField(Opcodes.ACC_PUBLIC, "needsResourceRefresh", "Z", null, null).visitEnd()
        original.visitField(Opcodes.ACC_PUBLIC, "allKeys", "Ljava/util/List;", null, null).visitEnd()
        original.koffee {
            method5(public, "onGuiClosed", void) {
                aload_0
                getfield(GameSettings::class, "needsResourceRefresh", boolean)
                ifeq(L["1"])
                aload_0
                getfield(GameSettings::class, "mc", Minecraft::class)
                invokevirtual(Minecraft::class, "scheduleResourcesRefresh", ListenableFuture::class)
                pop
                aload_0
                iconst_0
                putfield(GameSettings::class, "needsResourceRefresh", boolean)
                +L["1"]
                _return
                maxStack = 2
                maxLocals = 1
            }
        }

        original.methods.forEach {
            when (it.name) {
                "<init>" -> {
                    val (createAllKeys) = assembleBlock {
                        aload_0
                        new(ArrayList::class)
                        dup
                        invokespecial(ArrayList::class, "<init>", void)
                        putfield(GameSettings::class, "allKeys", List::class)
                    }

                    if (it.desc == "(Lnet/minecraft/client/Minecraft;Ljava/io/File;)V") {
                        for (insn in it.instructions.iterator()) {

                        }
                    }
                }

                "<clinit>" -> {
                    val (createNewArrayItem) = assembleBlock {
                        dup
                        iconst_3
                        ldc("options.particles.disabled")
                        aastore
                    }

                    for (insn in it.instructions.iterator()) {
                        val (changeNode) = assembleBlock {
                            iconst_4
                        }

                        if (insn is LdcInsnNode && insn.cst == "options.particles.minimal") {
                            it.instructions.insertBefore(insn.next?.next, createNewArrayItem)
                        }

                        val next = insn.next?.next?.next?.next
                        if (insn is InsnNode && insn.opcode == Opcodes.ICONST_3 && next is LdcInsnNode && next.cst == "options.particles.all") {
                            it.instructions.insertBefore(insn, changeNode)
                            it.instructions.remove(insn)
                        }
                    }
                }

                "setOptionFloatValue" -> {
                    val (insertBoolean) = assembleBlock {
                        aload_0
                        iconst_1
                        putfield(GameSettings::class, "needsResourceRefresh", boolean)
                        _return
                    }

                    for (insn in it.instructions.iterator()) {
                        if (insn is MethodInsnNode && insn.owner == "net/minecraft/client/Minecraft" &&
                            insn.name == "scheduleResourcesRefresh" && insn.desc == "()Lcom/google/common/util/concurrent/ListenableFuture;"
                        ) {
                            it.instructions.insertBefore(insn.previous?.previous, insertBoolean)
                        }
                    }

                    it.instructions.insert(
                        MethodInsnNode(
                            Opcodes.INVOKEVIRTUAL,
                            "net/minecraft/client/renderer/texture/TextureMap",
                            "setBlurMipmapDirect",
                            "(ZZ)V",
                            false
                        ), insertBoolean
                    )
                }

                "setOptionValue" -> {
                    val (changeNode) = assembleBlock {
                        iconst_4
                    }

                    for (insn in it.instructions.iterator()) {
                        if (insn is InsnNode && insn.opcode == Opcodes.ICONST_3 &&
                            insn.next?.next is FieldInsnNode && (insn.next
                                ?.next as FieldInsnNode).name == "particleSetting"
                        ) {
                            it.instructions.insertBefore(insn, changeNode)
                            it.instructions.remove(insn)
                        }
                    }
                }
            }
        }

        return original
    }
}
